add out variant of cudnn fused conv

makslevental opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #61824
* #61821
* __->__ #61802
* #61801
* #61800

resubmitting because the last one was unrecoverable due to making changes incorrectly in the stack

Differential Revision: [D29812513](https://our.internmc.facebook.com/intern/diff/D29812513)

makslevental committed:
add out variant of cudnn fused conv

[ghstack-poisoned]

desertfire reviewed (approved):
LGTM

Commit:
Update on "add out variant of cudnn fused conv"


resubmitting because the last one was unrecoverable due to making changes incorrectly in the stack

[ghstack-poisoned]

makslevental commented:
@makslevental has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29812513).

ngimel commented on aten/src/ATen/native/cudnn/ConvShared.cpp:
```diff
@@ -565,6 +555,23 @@ Tensor cudnn_convolution_add_relu(
+    int64_t groups,
+    Tensor& output_t
+    ) {
   if (output_t.numel() == 0) {
```
`cudnn_convolution_add_relu_out` is a public API, which means that users can send arbitrarily-sized `output_t` arguments to it. 0-sized `output_t` doesn't mean early exit, incorrectly-sized `output_t` has to either error out or be resized. We are trying to deprecate resizing of incorrectly sized `out` args, except resizing 0-element to correct size, so erroring out is better I'd say.

makslevental replied:
@ngimel sounds good. suggestion for which error to use?

ngimel replied:
You can use `at::native::resize_output` which will throw warning for now if output is of the wrong size, then at some point we'll switch it to error.

makslevental replied:
@ngimel all set. if this is correct can you please approve.

makslevental committed:
Update on "add out variant of cudnn fused conv"


resubmitting because the last one was unrecoverable due to making changes incorrectly in the stack

Differential Revision: [D29812513](https://our.internmc.facebook.com/intern/diff/D29812513)

[ghstack-poisoned]

makslevental commented:
@makslevental has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29812513).

makslevental commented:
@makslevental has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29812513).

makslevental committed:
Update on "add out variant of cudnn fused conv"


resubmitting because the last one was unrecoverable due to making changes incorrectly in the stack

Differential Revision: [D29812513](https://our.internmc.facebook.com/intern/diff/D29812513)

[ghstack-poisoned]
